src/component/I18nRoutingProvider.js

```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import I18nRoutingContext from './I18nRoutingContext';
+
+class I18nRoutingProvider extends React.Component {
+  state = {
+    locale: null,
+  };
+
+  constructor(props) {
+    super(props);
+
+    const location = props.history ? props.history.location : {};
+
+    this.state = {
+      locale: props.localeFromPath({
+        ...location,
+        hostname: typeof window !== 'undefined' ? window.location.hostname : '',
+      }),
+      translatedRoutes: props.defaultTranslatedRoutes || {},
+      setTranslatedRoutes: this.setTranslatedRoutes.bind(this),
+    };
+  }
+
+  static getDerivedStateFromProps(props, state) {
```

Comment:
True, maybe we can move it to render, I just want to pass formatIntlRoute to the context. 

```diff
-  static getDerivedStateFromProps(props, state) {
-    return {
-      ...state,
-      formatIntlRoute: props.formatIntlRoute,
-    };
-  }
   render() {
     return (
-      <I18nRoutingContext.Provider value={this.state}>
+      <I18nRoutingContext.Provider value={{formatIntlRoute: this.props.formatIntlRoute, ...this.state}}>
         {this.props.children}
       </I18nRoutingContext.Provider>
     );
   }
```

Reply:
`{formatIntlRoute: this.props.formatIntlRoute, ...this.state}`
This will force a re-render on the Consumer since the passed value will always be a new object.

Reply:
Any better option?

Reply:
As the goal is to pass down the provided `formatIntlRoute` prop from the Provider to the Consumer and this prop should only be set on the first mount, we could assign it to the Provider's state inside the constructor. 

If we really want to update our Consumers if the Provider is re-rendered with a new prop value I think we should go back to the previous implementation.

Reply:
I think that those things are micro-optimizations that they can be introduce after this PR if we see that this implementation is generating bottlenecks, otherwise it does not have any sense. IMHO

Reply:
I agree with @benatespina. As long as we keep the public API untouched with further optimizations I would go with this PR as it is